src/app/hooks/useTypedSearchParam.ts

```diff
@@ -0,0 +1,44 @@
+import { useSearchParams } from 'react-router-dom'
+
+export type UrlSettingOptions = {
+  /**
+   * Should we replace the current state in the history, instead of appending a new state?
+   *
+   * Default is false
+   */
+  replace?: boolean
+
+  /**
+   * Should we block scrolling to the top?
+   *
+   * Default is true
+   */
+  preventScrollReset?: boolean
+
+  /**
+   * Should we delete some other params while setting this one?
+   */
+  deleteParams?: string[]
+}
+
+type SetterFunction<T> = (value: T, options?: UrlSettingOptions) => void
+
+export function useTypedSearchParam<T = string>(
+  paramName: string,
+  defaultValue: T,
+  defaultSetterOptions: UrlSettingOptions = {},
+): [T, SetterFunction<T>] {
```

Comment:
Instead of `T = string`, lets do something like https://github.com/oasisprotocol/explorer/blob/61c87ee24567997f684ddf23314492b1e318e436/src/app/utils/storage.ts#L18 in the future